Networking/Networking/Requests/AuthenticatedRequest.swift

```diff
@@ -8,29 +8,43 @@ enum AuthenticatedRequestError: Error {
 /// Wraps up a URLRequestConvertible Instance, and injects the Credentials + `Settings.userAgent` whenever the actual Request is required.
 ///
 struct AuthenticatedRequest: URLRequestConvertible {
-
-    /// WordPress.com Credentials.
+    /// Authenticated Request
     ///
-    let credentials: Credentials
+    let request: URLRequest
 
-    /// Request that should be authenticated.
-    ///
-    let request: URLRequestConvertible
+    init(authToken: String, request: URLRequest) {
+        var authenticated = request
 
+        authenticated.setValue("Bearer " + authToken, forHTTPHeaderField: "Authorization")
+        authenticated.setValue("application/json", forHTTPHeaderField: "Accept")
+        authenticated.setValue(UserAgent.defaultUserAgent, forHTTPHeaderField: "User-Agent")
 
-    /// Returns the Wrapped Request, but with a WordPress.com Bearer Token set up.
-    ///
-    func asURLRequest() throws -> URLRequest {
-        guard case let .wpcom(_, authToken, _) = credentials else {
-            throw AuthenticatedRequestError.invalidCredentials
-        }
+        self.request = authenticated
+    }
 
-        var authenticated = try request.asURLRequest()
+    init(applicationPassword: ApplicationPassword, request: URLRequest) {
+        var authenticated = request
 
-        authenticated.setValue("Bearer " + authToken, forHTTPHeaderField: "Authorization")
         authenticated.setValue("application/json", forHTTPHeaderField: "Accept")
         authenticated.setValue(UserAgent.defaultUserAgent, forHTTPHeaderField: "User-Agent")
 
-        return authenticated
+        let username = applicationPassword.wpOrgUsername
+        let password = applicationPassword.password.secretValue
+        let loginString = "\(username):\(password)"
+
+        if let loginData = loginString.data(using: .utf8) {
+            let base64LoginString = loginData.base64EncodedString()
+            authenticated.setValue("Basic \(base64LoginString)", forHTTPHeaderField: "Authorization")
+        }
+
+        // Cookies from `CookieNonceAuthenticator` should be skipped
+        authenticated.httpShouldHandleCookies = false
+        self.request = authenticated
```

Comment:
After understanding the code a bit more, it looks like the REST request-related logic is in two `URLRequestConvertible` places - `AuthenticatedRequest` (authenticating a REST request) and `RESTRequest` (generating a URL for a REST request). Since these two have to go hand in hand, I'm afraid at some point we're going to mistakenly use `AuthenticatedRequest`'s WPCOM authentication with a `RESTRequest` or vice versa. How about separating `AuthenticatedRequest` into `AuthenticatedDotcomRequest` and `AuthenticatedRESTRequest` so that it's clear which one to use with each type of request?

Reply:
Done in e30bfd550c4f612e268335e043dd356ea436153a